### ISSUE TITLE
Kindle Scribe - wrong network selected when in new place

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#12639]
* KOReader version: 2024.07-250-gce7cff36
* Device: Kindle Scribe (5.16.21 HF)

#### Issue and Steps to reproduce
So, this issue (part 3) is real, and happened to me again: https://github.com/koreader/koreader/pull/12056#issuecomment-2203442629

1) I was at home, connected to home Wi-Fi network.
2) Then I went to work. At work, I have `eduroam` setup on Kindle.
3) However, once I turn on Wi-Fi with a gesture, two things go wrong:
a) Kindle says "Connecting to HOME_NETWORK"
b) Kindle then proceeds to connect to a new password-less network (starting with "Vive-", a few of them visible where I am in the office).

I am sure that this network is wrong because I wiped one such network before following these steps: https://www.mobileread.com/forums/showpost.php?p=4457977&postcount=2

##### `crash.log` (if applicable)
I don't think verbose logging catches the details of new networks it actually connects to.
I see in the log (right after I did the "W" multi-swipe to enable Wi-Fi):
```
10/14/24-10:58:39 DEBUG multiswipe southeast northeast southeast northeast 
10/14/24-10:58:39 DEBUG Contact:handleSwipe: swipe east 519.593110039 detected 
10/14/24-10:58:39 DEBUG AutoSuspend: onInputEvent 
10/14/24-10:58:39 DEBUG ImageWidget: _render'ing resources/icons/mdlight/notice-info.svg 77 77 
10/14/24-10:58:39 DEBUG ImageWidget: initial offsets 0 0 
10/14/24-10:58:39 DEBUG show widget: table: 0xb02fa330 
10/14/24-10:58:39 DEBUG setDirty nil from widget table: 0xb02fa330 w/ NO region; dithering: nil 
10/14/24-10:58:39 DEBUG setDirty via a func from widget table: 0xb02fa330 
10/14/24-10:58:39 DEBUG painting widget: table: 0xb02fa330 
10/14/24-10:58:39 DEBUG blitFrom 261 1201 0 0 77 77 
10/14/24-10:58:39 DEBUG _refresh: Enqueued ui update for region 248 1188 1363 103 dithering: false 
10/14/24-10:58:39 DEBUG triggering refresh {
  dither = false,
  mode = "ui",
  region = "1363x103+248+1188"
} --[[table: 0xb46cd948]] 
10/14/24-10:58:39 DEBUG refresh: ui-mode 248 1188 1363 103 false 
10/14/24-10:58:39 DEBUG refresh: wait for submission of (previous) marker 146 
10/14/24-10:58:39 DEBUG mxc_update: 1363x103 region @ (248, 1188) with marker 147 (WFM: 257 & UPD: 0) 
10/14/24-10:58:39 DEBUG AutoSuspend: onNetworkConnecting 
10/14/24-10:58:39 DEBUG ImageWidget: _render'ing resources/icons/mdlight/notice-info.svg 77 77 
10/14/24-10:58:39 DEBUG ImageWidget: initial offsets 0 0 
10/14/24-10:58:39 DEBUG show widget: table: 0xb4b097f8 
10/14/24-10:58:39 DEBUG setDirty nil from widget table: 0xb4b097f8 w/ NO region; dithering: nil 
10/14/24-10:58:39 DEBUG setDirty via a func from widget table: 0xb4b097f8 
10/14/24-10:58:39 DEBUG painting widget: table: 0xb4b097f8 
10/14/24-10:58:39 DEBUG blitFrom 261 1201 0 0 77 77 
10/14/24-10:58:39 DEBUG _refresh: Enqueued ui update for region 248 1188 1363 103 dithering: false 
10/14/24-10:58:39 DEBUG triggering refresh {
  dither = false,
  mode = "ui",
  region = "1363x103+248+1188"
} --[[table: 0xb45d48b8]] 
10/14/24-10:58:39 DEBUG refresh: ui-mode 248 1188 1363 103 false 
10/14/24-10:58:39 DEBUG refresh: wait for submission of (previous) marker 147 
10/14/24-10:58:39 DEBUG mxc_update: 1363x103 region @ (248, 1188) with marker 148 (WFM: 257 & UPD: 0) 
10/14/24-10:58:39 DEBUG kindleScanThenGetResults: Wi-Fi scan took 0 seconds 
10/14/24-10:58:39 DEBUG close widget: table: 0xb4b097f8 
10/14/24-10:58:39 DEBUG setDirty via a func from widget nil 
10/14/24-10:58:39 DEBUG Lower widget ReaderUI was dithered, honoring the dithering hint 
10/14/24-10:58:39 DEBUG Lower widget ReaderUI covers the full screen 
10/14/24-10:58:39 DEBUG setDirty nil from widget ReaderUI w/ NO region; dithering: true 
10/14/24-10:58:39 DEBUG setDirty nil from widget table: 0xb02fa330 w/ NO region; dithering: nil 
10/14/24-10:58:39 DEBUG NetworkMgr: Attempting to authenticate on preferred network BHNTG1682G5527 
10/14/24-10:58:39 DEBUG ImageWidget: _render'ing resources/icons/mdlight/notice-info.svg 77 77 
10/14/24-10:58:39 DEBUG ImageWidget: initial offsets 0 0 
10/14/24-10:58:39 DEBUG show widget: table: 0xb027e2c8 
10/14/24-10:58:39 DEBUG setDirty nil from widget table: 0xb027e2c8 w/ NO region; dithering: nil 
10/14/24-10:58:39 DEBUG setDirty via a func from widget table: 0xb027e2c8 
10/14/24-10:58:39 DEBUG NetworkMgr: Connected to network BHNTG1682G5527  
```
but it actually connected to this "Vive-..." network.

What can I do to have better logs?

**Comment by mergen3107** [comment]
Important thing to add:
Kindle itself never connected to new "Vive-" networks itself.
Only to ones I connected to before.

So I am quite confident that it is KOReader doing something wrong.

**Comment by NiLuJe** [comment]
Again, unless you explicitly get to the AP list and tap on one, we don't do *anything* besides requesting *a* connection: wifid is in charge of doing whatever the hell it thinks ought to be happening automatically ;).

You can check/follow the system logs for (possibly) more details about all that. (`showlog`).

**Comment by mergen3107** [comment]
`wifid is in charge of doing whatever the hell it thinks ought to be happening automatically`
Huh, what else I found weird is that after KOReader picks up this "Vive-" network, I exit to Kindle UI and it shows that as if it is not connected.

Thanks! I'll try to see with showlog (after I enable it back :D)

**Comment by NiLuJe** [comment]
> I exit to Kindle UI and it shows that as if it is not connected.

And is it, really?

(That's... a potentially interesting detail, and something you might want to track in the syslogs, too).

**Comment by mergen3107** [comment]
Yes, it is not connected when I am in Kindle UI.

OTOH, KOReader can even tell me device's IP address while it is on this "Vive-" network.

I cannot, however, test this network - I don't know if it has access to internet at all. Might be some hotspot for our OIT department (they are in the same building as I am)

**Comment by mergen3107** [comment]
Yesterday I reverted back to FW 5.16.2.0.1 on Kindle Scribe (not that it matters here).

Now I found myself stuck in KOReader with this "Vive-" network.

Here is what happened:
1) After downgraded, Kindle has not yet been connected to eduroam. I believe, it was still in the list of `wpa_cli list_networks`.
2) I turned on Wi-Fi in KOReader (nightly 251). It connected to this "Vive-" network (in Network info).
3) I exited KOReader. Wi-Fi was on, but it showed as if it didn't connect to anything.
4) I hit eduroam - after about 1 minutes, it complained that I need to set it up again (usual eduroam stuff).
5) I set it up, connected. Everything worked (Kindle store, sftp).
6) I opened KOReader. It reported that it is connected to eduroam (in Network info).
7) Now, if I turn off Wi-Fi and back on in KOReader - it will connected to this "Vive-" network again!
I am stuck, because KOReader keeps connecting to it.

I long-pressed "Wi-Fi connection" while Wi-Fi was off, I can see (finally - I didn't know how to summon it manually before :D) the screen with networks. Here I even did "edit" to Forget this Vive-network, but it didn't help.

What am I doing wrong?

**Comment by mergen3107** [comment]
After all this, in that same network list screen I edited eduroam and entered password.
KOReader has a positive checkbox next to "Wi-Fi connection", but progress bar widget doesn't reflect that, Network info says it is not connected, and Kindle UI also shows that Airplane mode is on (i.e., Wi-Fi off). 

TLDR; I cannot connect to eduroam from within KOReader anymore.

I'll try to get some logs after I connected to eduroam from Kindle UI again.

**Comment by mergen3107** [comment]
```
10/16/24-15:27:38 DEBUG NetworkMgr: Attempting to authenticate on preferred network Vive-017d6a30 
10/16/24-15:27:38 DEBUG ImageWidget: _render'ing resources/icons/mdlight/notice-info.svg 77 77 
10/16/24-15:27:38 DEBUG ImageWidget: initial offsets 0 0 
10/16/24-15:27:38 DEBUG show widget: table: 0xb4b7efa0 
10/16/24-15:27:38 DEBUG setDirty nil from widget table: 0xb4b7efa0 w/ NO region; dithering: nil 
10/16/24-15:27:38 DEBUG setDirty via a func from widget table: 0xb4b7efa0 
10/16/24-15:27:38 DEBUG NetworkMgr: Connected to network Vive-017d6a30 
```
1) it was not connected to before at all (wiped from by wpa_cli!)
2) it is definitely not preferred at all, lol

Full log:
[crash_wrong_network2.txt](https://github.com/user-attachments/files/17402177/crash_wrong_network2.txt)

**Comment by mergen3107** [comment]
Also, what is this ubiquitous 
```
10/16/24-15:28:24 DEBUG NetworkMgr: interface is not operational yet
```
?

**Comment by NiLuJe** [comment]
> 1. wpa_cli

Don't play with wpa_supplicant manually on a Kindle.

Check wifid's stuff over lipcs (I don't know the exact syntax, we probably talked about it in the original PR).

> DEBUG NetworkMgr: interface is not operational yet

That's the connectivity check.

> Full log:

Probably unhelpful: I'm repeating myself, but we don't really control wifi on that platform ;o).

**Comment by mergen3107** [comment]
Here is everything from `lipc-probe -v -a` about `wifid`:
```
com.lab126.wifid
	rw 	Has	certificateData	[*NOT SHOWN*]
	w	Str	deleteCertificate
	rw	Has	cmIntfInfo	[*NOT SHOWN*]
	r	Str	signalStrength	[5/5]
	rw	Int	enable	[1]
	w	Int	hotSpotDBDownloadStatus
	r	Str	711	[... my network details ...]
	r	Str	macSecret	[***]
	w	Str	cmConnect
	w	Str	cmCheckConnection
	rw	Has	connMetadata	[*NOT SHOWN*]
	r	Int	cmIntfInUse	[1]
	r	Str	manufacturerCode	[***]
	rw	Has	installCertificate	[*NOT SHOWN*]
	w	Str	setRegDomCC
	r	Str	macAddress	[***]
	rw	Has	profileData	[*NOT SHOWN*]
	rw	Has	netConfig	[*NOT SHOWN*]
	w	Str	cmDisconnect
	rw	Str	logMask	[0x0fff0000]
	rw	Str	logLevel	[Current log level=info
(Possible transient levels: all, perf, debug[9-0], info, warn, error, crit, none)
(Possible persistent levels: p_all, p_perf, p_debug[0-9], p_info, p_warn, p_error, p_crit, p_none)]
	r	Int	scanListCount	[29]
	rw	Has	cmNWProperties	[*NOT SHOWN*]
	rw	Has	createProfile	[*NOT SHOWN*]
	r	Str	scanState	[idle]
	w	Str	cmConnMode
	r	Str	feelingLuckyProfile	[]
	w	Int	deleteProfile
	rw	Has	currentEssid	[*NOT SHOWN*]
	r	Str	cmState	[CONNECTED]
	rw	Has	createNetConfig	[*NOT SHOWN*]
	w	Str	scan
	rw	Has	netidValue	[*NOT SHOWN*]
	rw	Has	scanList	[*NOT SHOWN*]
	r	Int	profileCount	[6]
	w	Int	resetSavePasswordPreference
```

which one of those if helpful? :D 

Thanks! I'll go look into the system logs. I need to unfreeze logging capability first (didn't wanna do it since Amazon collects too much info, but that's the only way to know what's going on it seems).

**Comment by mergen3107** [comment]
Found it here:
https://github.com/koreader/koreader/pull/12056/files#diff-bc1f1d3fd8a1f49053e6f1cdde7fde96c9c790b85c5379ff4852739bcd05b7b7R26

How to access it via normal command line?
I can see with a Tab that `lipc-` has these:
```
[root@kindle root]# lipc-
lipc-daemon           lipc-get-prop-retry   lipc-hash-prop-retry  lipc-send-event       lipc-set-prop-retry
lipc-get-prop         lipc-hash-prop        lipc-probe            lipc-set-prop         lipc-wait-event
```
but `lipc-hash-prop com.lab126.wifid com.lab126.wifid profileData` doesn't really work (it just goes into a new line as if waiting for a command, but nothing happens).

![image](https://github.com/user-attachments/assets/3f8659f7-31b0-464a-b5f0-3e511d09b22f)

**Comment by mergen3107** [comment]
Alright, thanks again NiLuJe!

1) The problem is indeed with Kindle, or rather `eduroam` itself (or rather, an enterprise network lease). It seems that `eduroam` doesn't reliably work for other people too: https://www.reddit.com/r/kindlescribe/comments/14l8jbl/argh_still_cant_log_into_eduroam_reliably/

2) I got misled by accidents. I thought, it only happens in KOReader. However, a few times today (out of 25-30 reconnections by entering credentials again...) I was able to catch it in the Kindle UI.

3) Problem as I see (without logs) if following:
- eduroam provides a "lease" for some period of time, probably lasting less than a day. Not sure, because sometimes I am able to keep connecting to it day after day, sometime I can't reconnect within an hour.
- "Lease" expires. However, Kindle doesn't know how to deal with this nicely, and most probably falls back to the next available and recorded network.
- Ideally, it should say "Hey, cannot connect to this network anymore, enter credentials again". Instead, from KOReader it "connects" to "Vive-" network. I say in quotes, because when I actually try to connect to it in Kindle UI, it rejects it. KOReader, however, says it is connected to it and provides an IP address :D 
- When I hit this in Kindle UI, it just silently stays unconnected to any network. Now, if I hit `eduroam` in the Kindle UI network list again - after about a minute of thinking it says "Oh you need to set it up". I enter credentials again, and I am back on the network.
- I vaguely remember that I accidentally hit that "Vive-" network about half a year ago. So it probably falls back to that, over and over.

Feel free to add a tag "Not our bug" and "Won't Fix".
Even if I get proper logs, I don't think there is anything we can do if Kindle just doesn't know how to enterprise network. KOReader just follows what `currentEssid` is, which probably changes after "lease" expires.

I don't know, however, **how to delete that "Vive-" network for good**. It kept coming back after `wpa_cli` dances. Without it, I hope, KOReader will just plainly stay not connected if eduroam fails.
Any help is appreciated!

Thanks!

P.S. Sorry for being suspicious regarding KOReader at first :D I was just following my observations.

**Comment by NiLuJe** [comment]
I think we talked about how to poke at the saved networks list in the initial openlipclua/AP list PR?

**Comment by mergen3107** [comment]
You mean many years ago?

**Comment by mergen3107** [comment]
I enabled logs and did `showlog -f` via USB while playing with Wi-Fi.

Findings:
1) After Wi-Fi is turned on, `eduroam` connection is tried many times. Sometimes, it goes through, but most of the times it doesn't.
2) Erasing Wi-Fi profiles with `wpa_cli` and with methods described here () does not persist. The same profiles are back after I turn off Wi-Fi and back on.

"Vive-" network becomes a fallback after maybe 15 fails with eduroam.

I think my main problem is that I have a duplicate of eduroam. Kindle still tries to "connect" to a network named "eduroam", even though I manually did "Forget this network" in Kindle UI!

Time for a factory reset? I wonder if going between hf and non-hf versions did something wrong.

Here is the log after I turn on Wi-Fi and it fails about 15 times.
[eduroam.txt](https://github.com/user-attachments/files/17439902/eduroam.txt)

**Comment by Frenzie** [comment]
You also can't remove it in the Kindle interface? Does the device have a /etc/wpa_supplicant/wpa_supplicant file you can simply wipe to forget wifi connections?

**Comment by mergen3107** [comment]
Yes, once I hit it again, it asks me to input credential as if it is a new network :D 
In /etc/, I can only see `wpa_supplicant.conf` file. No folder/file called `wpa_supplicant`.
`wpa_supplicant.conf` points to `/var/run/wpa_supplicant`, but there is no such file.

There a file `/var/local/system/wifilocker/wifilocker.db`, but renaming it doesn't seem to do much.

**Comment by Frenzie** [comment]
> I can only see wpa_supplicant.conf file.

Yes sorry, that's the one I meant. ^_^ See https://linux.die.net/man/5/wpa_supplicant.conf

**Comment by mergen3107** [comment]
Thanks @Frenzie !

It is pretty much useless though :D 
Only has up to 10 lines, most of which look generic.

**Comment by Frenzie** [comment]
Does the Kindle use something fancier like NetworkManager perhaps?

**Comment by mergen3107** [comment]
It doesn't like it, there is no /etc/NetworkManager/ folder neither. 

I probably uses proprietary Amazon wpa :D

**Comment by NiLuJe** [comment]
> You mean many years ago?

Nope, in the Wi-Fi selector PR, but I can't seem to find it at a quick glance.

In any case, I vaguely recall that there was something to poke at the wifid list of saved networks :?

**Comment by mergen3107** [comment]
Ah, that :) 

Yep, I found all of it after that post. See my investigation above. 

TL;DR
Amazon's wifid implementation sucks :D 
And doesn't like enterprise networks as much (but thanks they work _some_ way, wasn't the case in Kindle Touch)

**Comment by mergen3107** [comment]
FTR, Wi-Fi selector was added in this PR:

https://github.com/koreader/koreader/pull/12056
